Add crooked pipe

alexrlongne opened the pull request:
## Background

* We'd like to compare the moment method and IMC on a common radiation test problem

## Description of Changes

* Add pgen input for crooked pipe with geometric density and temperature specifications
* Modify pgen.hpp to include crooked pipe case
* Modify params.yaml to include crooked pipe defaults

## Checklist

- [ ] New features are documented
- [ ] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files

alexrlongne committed:
Add crooked pipe input option to pgen

+ Add pgen/crooked_pipe.hpp with specific geometric crooked pipe
 density and source parameters
+ Update list in params.yaml to include crooked pipe input options
 and defaults
+ Update if/else list in pgen.hpp to include crooked_pipe

alexrlongne committed:
+ Fix error in pgen crooked pipe file
+ Add crooked pipe input file

RyanWollaeger commented on src/pgen/crooked_pipe.hpp:
```diff
@@ -0,0 +1,162 @@
+  // * density of thin regions using rho_thin
+  // * density of thick regions using rho_thick
+  // * initial material and radiation temperature of all cell using t_init
+  if (do_imc) {
```
Suppose for moments, this if-check could be moved to be just around the `jaybenne::InitializeRadiation` invocation below.

alexrlongne replied:
It sounds like Adam was going to follow-up this PR with some changes for the moments version.

alexrlongne committed:
Fix a few issues in input and pgen

+ Disable gas (don't merge!)
+ Fix regions in inputs

alexrlongne committed:
More crooked fixes

+ Use std::array instead of std::vector in crooked pipe pgen
 to fix initialization behavior on the GPU (host code only functions
 were causing no initialization)
+ Add "update_fluxes" bool to gas field, default to on, turn off
 for radiation only test problems
+ Print when "update_fluxes" is set to false in artemis_utils.cpp
+ Update crooked pipe input to have update_fluxes off
+ Update jaybenne to pull in continuous absorption

alexrlongne committed:
+ Format new files
+ Add check to make sure crooked pipe is run with update_fluxes
 set to false

alexrlongne committed:
Appending diffs from CI

+ Revert to old analysis plot script

adamdempsey90 commented on inputs/radiation/crooked_pipe.in:
```diff
@@ -0,0 +1,103 @@
+<artemis>
```
```suggestion

//========================================================================================
// (C) (or copyright) 2026. Triad National Security, LLC. All rights reserved.
//
// This program was produced under U.S. Government contract 89233218CNA000001 for Los
// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
// in the program are reserved by Triad National Security, LLC, and the U.S. Department
// of Energy/National Nuclear Security Administration. The Government is granted for
// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
// license in this material to reproduce, prepare derivative works, distribute copies to
// the public, perform publicly and display publicly, and to permit others to do so.
//========================================================================================
<artemis>
```

adamdempsey90 commented on src/pgen/crooked_pipe.hpp:
```diff
@@ -0,0 +1,165 @@
+//========================================================================================
+// (C) (or copyright) 2023-2025. Triad National Security, LLC. All rights reserved.
```
```suggestion
// (C) (or copyright) 2026. Triad National Security, LLC. All rights reserved.
```

adamdempsey90 commented on src/pgen/crooked_pipe.hpp:
```diff
@@ -0,0 +1,165 @@
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);
+
+  std::array<std::array<double, 4>, 7> thick_regions = {
+      {{3.0, 4.0, -1.0, 1.0},
+       {-2.0, 2.5, -2.0, -0.5}, // extended xl to -2 for thick region above source
+       {-2.0, 2.5, 0.5, 2.0},   // extended xl to -2 for thick region above source
+       {4.5, 7.0, -2.0, -0.5},
+       {4.5, 7.0, 0.5, 2.0},
+       {2.5, 4.5, -2.0, -1.5},
+       {2.5, 4.5, 1.5, 2.5}}};
+
+  std::array<std::array<double, 4>, 1> thin_source_regions = {{{-2.0, 0.0, -0.5, 0.5}}};
```
These are technically not portable in C++17, but I will replace them in my follow on

adamdempsey90 reviewed (approved):
I will add the copyrights, run tests, and merge. Then, I'll make another MR with my modifications

adamdempsey90 committed:
Update inputs/radiation/crooked_pipe.in

Add copyright

adamdempsey90 committed:
Update src/pgen/crooked_pipe.hpp

Update copyright

adamdempsey90 committed:
Merge branch 'develop' into add_crooked_pipe
